browser/ui/views/playlist/thumbnail_provider.cc

```diff
@@ -0,0 +1,137 @@
+  return item->thumbnail_path.is_valid() &&
+         item->thumbnail_path != item->thumbnail_source;
+}
+
+}  // namespace
+
+ThumbnailProvider::ThumbnailProvider(playlist::PlaylistService* service)
+    : service_(service) {
+  CHECK(service);
+}
+
+ThumbnailProvider::ThumbnailProvider(playlist::PlaylistTabHelper* tab_helper)
+    : ThumbnailProvider(playlist::PlaylistServiceFactory::GetForBrowserContext(
+          tab_helper->web_contents()->GetBrowserContext())) {}
+
+ThumbnailProvider::~ThumbnailProvider() = default;
+
+void ThumbnailProvider::GetThumbnail(
+    const playlist::mojom::PlaylistItemPtr& item,
+    base::OnceCallback<void(const gfx::Image&)> callback) {
+  DVLOG(2) << __FUNCTION__;
+  auto& in_memory_cache = GetInMemoryCache(service_);
+  if (auto iter = in_memory_cache.Get(item->id);
+      iter != in_memory_cache.end()) {
+    std::move(callback).Run(iter->second);
+    return;
+  }
+
+  if (IsItemThumbnailCached(item)) {
+    if (base::FilePath thumbnail_path;
+        net::FileURLToFilePath(item->thumbnail_path, &thumbnail_path)) {
+      base::ThreadPool::PostTaskAndReplyWithResult(
+          FROM_HERE, base::MayBlock(),
+          base::BindOnce(
+              [](base::FilePath path) {
+                std::string raw_data;
+                if (base::ReadFileToString(path, &raw_data)) {
+                  return gfx::Image::CreateFrom1xPNGBytes(
+                      reinterpret_cast<unsigned char*>(raw_data.data()),
+                      raw_data.size());
+                }
+
+                VLOG(2) << __FUNCTION__ << " Failed to read " << path;
+                return gfx::Image();
+              },
+              thumbnail_path),
+          base::BindOnce(&ThumbnailProvider::OnGotThumbnail,
+                         weak_ptr_factory_.GetWeakPtr(), item->id,
+                         std::move(callback)));
+      return;
+    }
+  }
+
+  if (!item->thumbnail_source.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  service_->DownloadThumbnail(item->thumbnail_source,
+                              base::BindOnce(&ThumbnailProvider::OnGotThumbnail,
+                                             weak_ptr_factory_.GetWeakPtr(),
+                                             item->id, std::move(callback)));
+}
+
+void ThumbnailProvider::GetThumbnail(
+    const playlist::mojom::PlaylistPtr& list,
+    base::OnceCallback<void(const gfx::Image&)> callback) {
+  DVLOG(2) << __FUNCTION__;
+  if (!list->id.has_value()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  if (list->id == playlist::kDefaultPlaylistID) {
+    // If list is default folder, return default thumbnail image
+    // TODO
+    std::move(callback).Run({});
+    return;
+  }
+
+  // Else, find the first cached thumbnail.
+  auto iter = base::ranges::find_if(list->items, &IsItemThumbnailCached);
+  if (iter == list->items.end()) {
+    std::move(callback).Run({});
+    return;
+  }
```

Comment:
Curious this is what we want?
It seems give empty when all items doesn't have cached thumbnail in the list.
I thought it'll go ask network but doesn't.

Reply:
Yes, I think so. We don't show any images for playlist playlist when there's no thumbnail cached to local disk.

Reply:
Ah... this method is called for displaying target folder's thumbnail.
Maybe most of time it will give thumbnail but it could not have if thumbnail is not fetched yet for any reason?
Then, just using default thumbnail. am i right?

Reply:
Yup, you're right!